VisionOnEdge/ui/src/reducer/labelingPage.ts

```diff
@@ -0,0 +1,75 @@
+import {
+  UPDATE_ANNOTATION,
+  REQUEST_ANNOTATION_FAILURE,
+  REQUEST_ANNOTATION_SUCCESS,
+  CREATE_ANNOTATION,
+  UPDATE_CREATING_ANNOTATION,
+  FINISH_CREATING_ANNOTATION,
+  BoxObj,
+  REMOVE_ANNOTATION,
+  AnnotationAction,
+} from '../actions/labelingPage';
+import { initialState, LabelingPageState } from '../State';
+import { AnnotationState } from '../components/LabelingPage/types';
+
+const labelingPageStateReducer = (
+  state = initialState.labelingPageState,
+  action: AnnotationAction,
+): LabelingPageState => {
+  const newState = state;
+  switch (action.type) {
+    case REQUEST_ANNOTATION_SUCCESS:
+      newState.annotations = action.payload.annotations;
+      break;
+    case REQUEST_ANNOTATION_FAILURE:
+      break;
+    case CREATE_ANNOTATION:
+      newState.annotations.push(action.payload.annotation);
+      break;
+    case UPDATE_CREATING_ANNOTATION:
+      newState.annotations[newState.annotations.length - 1] = action.payload.updater(
+        newState.annotations[newState.annotations.length - 1],
+      );
+      newState.annotations = [...newState.annotations];
+
+      break;
+    case FINISH_CREATING_ANNOTATION:
+      {
+        const creatingAnnotation = newState.annotations[newState.annotations.length - 1];
+        if (creatingAnnotation.annotationState === AnnotationState.P1Added) {
+          if (
+            (creatingAnnotation.label.x1 | 0) === (creatingAnnotation.label.x2 | 0) &&
+            (creatingAnnotation.label.y1 | 0) === (creatingAnnotation.label.y2 | 0)
```

Comment:
Why using bitwise or here?

Reply:
Just let the values be integers before compare, we can use ```Math.round``` here instead.

Reply:
I see. I think using bitwise or should be fine. Thanks for the reply.